bee-message/src/address/mod.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+mod bls;
+mod ed25519;
+
+pub use bls::{BlsAddress, BLS_ADDRESS_LENGTH};
+pub use ed25519::{Ed25519Address, ED25519_ADDRESS_LENGTH};
+
+use crate::{error::ValidationError, signature::SignatureUnlock};
+
+use bee_packable::Packable;
+
+use bech32::{self, FromBase32, ToBase32, Variant};
+
+use alloc::{str::FromStr, string::String, vec::Vec};
+use core::convert::TryFrom;
+
+/// A generic address supporting different address kinds.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Packable)]
+#[cfg_attr(
+    feature = "enable-serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(tag = "type", content = "data")
+)]
+#[packable(tag_type = u8)]
+pub enum Address {
+    /// An Ed25519 address.
+    #[packable(tag = 0)]
+    Ed25519(Ed25519Address),
+    /// A BLS address.
+    #[packable(tag = 1)]
+    Bls(BlsAddress),
+}
+
+impl Address {
+    /// Returns the address kind of an `Address`.
+    pub fn kind(&self) -> u8 {
+        match self {
+            Self::Ed25519(_) => Ed25519Address::KIND,
+            Self::Bls(_) => BlsAddress::KIND,
+        }
+    }
+
+    /// Tries to create an `Address` from a Bech32 encoded string.
+    pub fn try_from_bech32(addr: &str) -> Result<Self, ValidationError> {
```

Comment:
currently bech32 addresses are represented as `&str`. 
should we maybe introduce a wrapper type for more type safety?
i know we had this in Bee before but decided against it, but isn't more nice to work with that wrapper type?

Reply:
any opinions on this @thibault-martinez ?

Reply:
Introduced the wrapper type `Bech32Address` 🙂